File: src/2025-02-24-deepclaude-rust/cpp/exercises/practice/sublist/sublist.cpp
#include "sublist.h"

namespace sublist {

namespace {

bool is_sublist(const std::vector<int>& sub, const std::vector<int>& super) {
    if (sub.size() > super.size()) {
        return false;
    }
    if (sub.empty()) {
        return true;
    }
    const size_t max_start = super.size() - sub.size();
    for (size_t i = 0; i <= max_start; ++i) {
        bool match = true;
        for (size_t j = 0; j < sub.size(); ++j) {
            if (super[i + j] != sub[j]) {
                match = false;
                break;
            }
        }
        if (match) {
            return true;
        }
    }
    return false;
}

} // namespace

List_comparison sublist(const std::vector<int>& a, const std::vector<int>& b) {
    if (a == b) {
        return List_comparison::equal;
    }
    if (a.empty()) {
        return List_comparison::sublist;
    }
    if (b.empty()) {
        return List_comparison::superlist;
    }
    bool a_is_sublist = is_sublist(a, b);
    bool b_is_sublist = is_sublist(b, a);
    if (a_is_sublist) {
        return List_comparison::sublist;
    }
    if (b_is_sublist) {
        return List_comparison::superlist;
    }
    return List_comparison::unequal;
}

}  // namespace sublist

File: src/2025-02-24-deepclaude-rust/go/exercises/practice/sublist/sublist.go
package sublist

// Relation type is defined in relations.go file.

func Sublist(l1, l2 []int) Relation {
	if equal(l1, l2) {
		return RelationEqual
	}
	if len(l1) == 0 {
		return RelationSublist
	}
	if len(l2) == 0 {
		return RelationSuperlist
	}
	if isSublist(l1, l2) {
		return RelationSublist
	}
	if isSublist(l2, l1) {
		return RelationSuperlist
	}
	return RelationUnequal
}

func equal(a, b []int) bool {
	if len(a) != len(b) {
		return false
	}
	for i, v := range a {
		if v != b[i] {
			return false
		}
	}
	return true
}

func isSublist(a, b []int) bool {
	if len(a) > len(b) {
		return false
	}
	for i := 0; i <= len(b)-len(a); i++ {
		match := true
		for j := 0; j < len(a); j++ {
			if b[i+j] != a[j] {
				match = false
				break
			}
		}
		if match {
			return true
		}
	}
	return false
}
